cpp/src/arrow/dataset/partition.h

```diff
@@ -106,6 +106,8 @@ struct ARROW_DS_EXPORT KeyValuePartitioningOptions {
   /// After splitting a path into components, decode the path components
   /// before parsing according to this scheme.
   SegmentEncoding segment_encoding = SegmentEncoding::Uri;
+  // Should the key be decoded according to the scheme above too
+  bool decode_key = false;
```

Comment:
Is there actually a reason to make this configurable? Is there a situation where we would want to leave special characters unescaped? @lidavidm

Reply:
Should I default decode_key to true or remove it entirely so that the key is always decoded?

Reply:
@zijie0 has a minimal reproducible example at: https://github.com/delta-io/delta-rs/issues/495#issue-1051108818. I would also prefer we do decode by default. If this could lead to breaking changes, then it makes sense to make it configurable, otherwise, it would be good to remove the option entirely.

Reply:
In that case I think let's remove the option entirely.

Reply:
Done, should be ready to be merged.